src/skadi/sample.cc

```diff
@@ -460,11 +447,13 @@ template <class coord_t> double sample::get(const coord_t& coord, tile_data& til
 
   // the caller can pass a cached tile, so we only fetch one if its not the one they already have
   if (index != tile.get_index()) {
-    // get the proper source of the data
-    tile = cache_->source(index);
-  }
-  if (!tile) {
-    return NO_DATA_VALUE;
+    if (!(tile = cache_->source(index))) {
```

Comment:
I think there's still a race between the following calls
1) `cache_source(index) ->   auto& item = cache[index];`
2)  `fetch(index) -> store -> cache_->insert -> cache[pos].init(..)`

2nd call is safe(as you lock `cache_lck` on `cache_->insert`), however 1st call is not under mutex

```suggestion
    {
      std::lock_guard<std::mutex> guard(cache_lck);
      tile = cache_->source(index);
    }
    if (!tile) {
```

Reply:
reason i didn't place lock there:
![threads](https://user-images.githubusercontent.com/7621906/148700600-c4426aea-02ad-4d04-8db3-95759a9ce944.jpg)

As i understood you are concerned of the case when we have **thread 4** and **thread 2**.
In my view we don't have race condition here because there is a lock  before insert method ( inside store method ) which limits access to this method to a single thread. This means that even if **thread 4** would get nullptr at cache[index] ( inside source method ) at the time when **thread 2** get value from the method get ( inside fetch method ) we would still successfully store the result with **thread 2** and initialise the cache with respective value. When the **thread 4** arrive we would have 2 possibilities: the store method is still locked or already unlocked but in both cases the **thread 4** will just contain the same data as the **thread 2** a moment ago and so we would not spoil the data.

Reply:
The concern is about how `cache_item_t` is accessed. 
From the writer side(`fetch` method):
1) lock the mutex
2) call `cache_item_t::init method`

From the reader side(`get` method):
1) get cache_item_t
2) if it is unitialized(nullptr) - call `item.init`([link](https://github.com/valhalla/valhalla/blob/4d02424fa5c60e1c65fd692b51bfcab80e9037a2/src/skadi/sample.cc#L334))

So it's possible that `cache_item_t::init` is called from the 2 threads at the same time. This is one of the possible races beside others(if compiler reorders instructions)

In theory, you could omit locking the mutex on reader side if `cache_item_t` was completely thread safe. But it is not